vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs

```diff
@@ -321,15 +334,82 @@ and
+                    member __.TargetFrameworkMoniker = ""
+                    member __.ProjectGuid =  project.Id.Id.ToString()
+                    member __.LoadTime = System.DateTime.Now
+                    member __.ProjectProvider = Some iProvideProjectSite
+                    member __.AssemblyReferences() = projectInfoManager.GetProjectInfo(project.FilePath) |> thrd
+                    member __.BuildErrorReporter with get () = errorReporter and 
+                                                      set (v) = errorReporter <- v
+                }
+
+        // TODO:   figure out why this is necessary
+        interface IVsHierarchy with
+            member __.SetSite(psp)                                    = hier.SetSite(psp)
+            member __.GetSite(psp)                                    = hier.GetSite(ref psp)
+            member __.QueryClose(pfCanClose)                          = hier.QueryClose(ref pfCanClose)
+            member __.Close()                                         = hier.Close()
+            member __.GetGuidProperty(itemid, propid, pguid)          = hier.GetGuidProperty(itemid, propid, ref pguid)
+            member __.SetGuidProperty(itemid, propid, rguid)          = hier.SetGuidProperty(itemid, propid, ref rguid)
+            member __.GetProperty(itemid, propid, pvar)               = hier.GetProperty(itemid, propid, ref pvar) 
+            member __.SetProperty(itemid, propid, var)                = hier.SetProperty(itemid, propid, var)
+            member __.GetNestedHierarchy(itemid, iidHierarchyNested, ppHierarchyNested, pitemidNested) = hier.GetNestedHierarchy(itemid, ref iidHierarchyNested, ref ppHierarchyNested, ref pitemidNested)
+            member __.GetCanonicalName(itemid, pbstrName)             = hier.GetCanonicalName(itemid, ref pbstrName)
+            member __.ParseCanonicalName(pszName, pitemid)            = hier.ParseCanonicalName(pszName, ref pitemid)
+            member __.Unused0()                                       = hier.Unused0()
+            member __.AdviseHierarchyEvents(pEventSink, pdwCookie)    = hier.AdviseHierarchyEvents(pEventSink, ref pdwCookie)
+            member __.UnadviseHierarchyEvents(dwCookie)               = hier.UnadviseHierarchyEvents(dwCookie)
+            member __.Unused1()                                       = hier.Unused1()
+            member __.Unused2()                                       = hier.Unused2()
+            member __.Unused3()                                       = hier.Unused3()
+            member __.Unused4()                                       = hier.Unused4()
+        }
+
+    member private this.OnProjectChanged(projectId:ProjectId, _newSolution:Solution) =
+        let project = this.Workspace.CurrentSolution.GetProject(projectId)
+        let siteProvider = this.ProvideProjectSiteProvider(this.Workspace, project)
+        this.SetupProjectFile(siteProvider, this.Workspace, "SetupNewTextView")
```

Comment:
Have you tested add/remove file, ad/remove reference and other changes to the project? 

I don't understand what happens as the project changes.  It seems that ``SetupProjectFile`` gets called multiple times with a different ``siteProvider`` each time.  But this feels wrong - ``SetupProjectFile`` is really only intended to be called once for each project.

Reply:
Please change ``"SetupNewTextView`` to ``"OnProjectChanged"`` (this text appears in FCS reactor trace diagnostics indicating the causal chain for operations int he FCS event queue)